Mentions API for interpretations and comments

adrianq opened the pull request:
- We added a new Dhis2PostgresSQLDialect in the original version of this PR. As we saw a new Dhis2PostgresDialect was added a few hours ago we merged our code with this one.
- Add generic json support for hibernate -> JsonUserType.java & JsonListUserType.java and extend for specific type MentionUserType
- Add Mention.java bean
- Modify interpretation and comments xml and bean to add mentions parameters.
- Add index for json property in TableAlteror.
- Add updatementions method in DefaultInterpretationService to extract mentions from text. It works when adding a new interpretation/comment and when editing a interpretation/comment
- Add test for mentions. Collateral effects:
   - We had to create MockUserService.
   - Modify h2 dialect and hibernate-default.properties to sort of handle jsonb
- Modify InterpretationController to support special API. Explanation:
   - api/interpretations?fields=*,comments[*]&filter=mentions.username:in:[android] -> just interpretations with mentions to 'android' user in the description
   - api/interpretations?fields=*,comments[*]&filter=comments.mentions.username:in:[android] -> just interpretations with mentions to 'android' user in the comments
   - api/interpretations?fields=*,comments[*]&filter=mentions:in:[android] -> interpretations with mentions to 'android' user in the comments and description. As we know, rootJunction was not working as expected I came up with this solution.

adrianq commented:
I have updated this branch. An internal message is now sent when a mention is done. Sharings are also updated adding read permission for the mentioned user if needed.

We have also created the PRs for the analytics tools (See https://github.com/dhis2/d2-analysis/pull/54). So these PRs close the loop at least for an initial version of the mentioning feature.

larshelge commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/hibernate/JsonUserType.java:
```diff
@@ -0,0 +1,206 @@
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
```
Have a look at "JsonBinaryType.java" in dhis-support-hibernate. Please use this type. We cannot have multiple JSON types in the system.

adrianq replied:
By the time we did this development there was no support for json types in Dhis2 so we implemented the support in a different way. We have just changed so we use JsonBinaryType...

larshelge commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/hibernate/JsonListUserType.java:
```diff
@@ -0,0 +1,101 @@
+    public JavaType createJavaType(ObjectMapper mapper) {
+        return mapper.getTypeFactory().constructCollectionType(List.class, returnedClass());
+    }
+
```
Have a look at "JsonBinaryType.java" in dhis-support-hibernate. Please use this type. We cannot have multiple JSON types in the system

larshelge commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/hibernate/MentionUserType.java:
```diff
@@ -0,0 +1,43 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.interpretation.Mention;
```
Have a look at "JsonBinaryType.java" in dhis-support-hibernate. Please use this type. We cannot have multiple JSON types in the system.

There is no point in creating new JSON user classes for each type. You can re-use the type we have, and define new Hibernate types in "dhis-suppprt-hibernate" -> UserTypes.hbm.xml.

larshelge commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,25 +200,97 @@ public void deleteInterpretation( Interpretation interpretation )
+                case EVENT_CHART:
+                    link += "/dhis-web-event-visualizer/index.html?id=" + interpretation.getChart().getUid() + "&interpretationid=" + interpretation.getUid();
+                    break;
+                default:
```
This will be fragile as it will break when we change or redeploy apps. Can we find a more robust way?

adrianq replied:
@larshelge I have got the idea from [this bit of code](https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/pushanalysis/DefaultPushAnalysisService.java#L340-L360) which is already part of master branch. Do you want me to extract these two pieces of code into a single method or do you have anything more sophisticated in mind? Any suggestion? I also had a look at LocalAppStorageService but doesn't seem the way to go...

larshelge commented on dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/interpretation/InterpretationServiceTest.java:
```diff
@@ -252,7 +312,8 @@ public void testLikeInterpretation()
     }
 
```
Can we add some additional tests?

adrianq replied:
We added some more tests for mentioning. If you have anything in particular you would like us to further testing please let us know...

larshelge replied:
Thank you.

larshelge commented on dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/mock/MockUserService.java:
```diff
@@ -0,0 +1,382 @@
+    @Override
+    public int getActiveUsersCount( Date since )
+    {
+        // TODO Auto-generated method stub
```
Please remove auto-generated comments.

adrianq replied:
Done!

larshelge commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/interpretation/Mention.java:
```diff
@@ -0,0 +1,75 @@
+    private String username;
+    
+    private Date created;
+    
```
Remove extra blank lines

larshelge commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/interpretation/Mention.java:
```diff
@@ -0,0 +1,75 @@
+@JacksonXmlRootElement( localName = "mentions", namespace = DxfNamespaces.DXF_2_0 )
+public class Mention implements Serializable
+{
+
```
Remove extra blank lines

larshelge commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/interpretation/MentionUtils.java:
```diff
@@ -0,0 +1,53 @@
+
+public class MentionUtils
+{
+    public static List<Mention> convertUsersToMentions(Set<User> users)
```
Codestyle, space around arg

larshelge commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/interpretation/MentionUtils.java:
```diff
@@ -0,0 +1,53 @@
+            mention.setUsername( user.getUsername() );
+            mentions.add( mention );
+        }
+        return (mentions.size() > 0) ? mentions : null;
```
I suggest do not return null from collection return type methods, rather empty collections, this avoid nullpointerexceptions downstream.

adrianq commented:
@larshelge I have taken care of all your comments/suggestions except the one regarding the logic inside 'sendNotifications'. I left a comment there as I am a bit unsure about what you have in mind...

larshelge commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,25 +200,97 @@ public void deleteInterpretation( Interpretation interpretation )
+    @Override
+    public void updateSharingForMentions( Interpretation interpretation, Set<User> users ){
+        for ( User user : users ) {
+            if (!aclService.canRead( user, interpretation.getObject() )) {
```
Codestyle, brackets on new line

larshelge commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,25 +200,97 @@ public void deleteInterpretation( Interpretation interpretation )
+            }
+
+            StringBuilder messageContent;
+            if (comment != null) {
```
Codestyle, brackets on new line

larshelge commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,25 +200,97 @@ public void deleteInterpretation( Interpretation interpretation )
+                    .append( comment.getText() );
+            }
+            else {
+                messageContent = new StringBuilder( "You were mentioned in the following interpretation: \n\n" )
```
Add internationalization using I18nManager to strings.

larshelge commented on dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/mock/MockUserService.java:
```diff
@@ -0,0 +1,338 @@
+    @Override
+    public void updateUser( User user )
+    {
+
```
Remove empty lines in all methods

larshelge commented on dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/mock/MockUserService.java:
```diff
@@ -0,0 +1,338 @@
+    @Override
+    public UserCredentials getUserCredentialsByUsername( String username )
+    {
+        for (User user: users) {
```
Codestyle, brackets

adrianq commented:
Addressed all the comments.
Re multilanguage, I added the terms as part of the global i18n file. There are not specificResources in dhis-services folder. Other option would be to send i18n as a parameter (it is being done like that in DefaultDataEntryFormService.java) but it would over-complicate the code quite a bit (interpretations api is being called from quite a few places).

larshelge commented on dhis-2/dhis-services/dhis-service-core/src/main/resources/i18n_global.properties:
```diff
@@ -1266,3 +1266,8 @@ import_settings=Import settings
 
+#-- Mentions -----------------------------------------------------------------#
+comment_mention_notification=You were mentioned in the following comment:
+interpretation_mention_notification=You were mentioned in the following interpretation:
```
Move the ":" out of the translation and into the template.

larshelge commented on dhis-2/dhis-services/dhis-service-core/src/main/resources/i18n_global.properties:
```diff
@@ -1266,3 +1266,8 @@ import_settings=Import settings
+comment_mention_notification=You were mentioned in the following comment:
+interpretation_mention_notification=You were mentioned in the following interpretation:
+go_to=Go to
+dhis2_mention=mentioned you in DHIS 2
```
Use "mentioned_you_in_dhis2" as key.

larshelge commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -35,23 +35,36 @@
 public class DefaultInterpretationService
-    implements InterpretationService
+    implements
+    InterpretationService
```
Same line

adrianq replied:
Just changed it - this was the result of applying DHIS2 code style format though

larshelge commented on dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/interpretation/InterpretationServiceTest.java:
```diff
@@ -55,10 +57,11 @@
 {
     @Autowired
-    private UserService userService;
+    private UserService _userService;
```
Can you just use "userService" here?

adrianq replied:
I don't think so. We need to inject userService in the test otherwise you get an error `java.lang.IllegalArgumentException: UserService must be injected in test`. If you look at `setUpTest` you will see that we are doing:
```
    @Override
    protected void setUpTest() throws Exception
    {
        userService = _userService;
    }
```

This is widely used in a lot of test in Dhis2. Look for `userService = _userService` and you will find more than 80 occurrences.

larshelge replied:
OK thanks.

larshelge commented on dhis-2/dhis-support/dhis-support-hibernate/src/main/resources/hibernate-default.properties:
```diff
@@ -13,7 +13,7 @@
 hibernate.connection.driver_class = org.h2.Driver
-hibernate.connection.url = jdbc:h2:mem:dhis2;DB_CLOSE_DELAY=-1;MVCC=TRUE;ALIAS_COLUMN_NAME=TRUE;DB_CLOSE_ON_EXIT=FALSE
+hibernate.connection.url = jdbc:h2:mem:dhis2;DB_CLOSE_DELAY=-1;MVCC=TRUE;ALIAS_COLUMN_NAME=TRUE;DB_CLOSE_ON_EXIT=FALSE;INIT=create domain if not exists jsonb as other;
 hibernate.connection.username = sa
```
What does this line do?

adrianq replied:
This line is needed so that H2 knows how to handle jsonb columns. Otherwise, when running InterpretationServiceTest you get an error for the methods that have to deal with the mentions column

larshelge replied:
OK thanks.

larshelge commented on dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/InterpretationController.java:
```diff
@@ -75,6 +89,56 @@
+                mentions.add( filterSplit[2] );
+                filterIterator.remove();
+            }
+        }
```
101-110: This is too much logic for a controller, can we move to InterpretationService or to MentionUtils?

adrianq replied:
Done here https://github.com/dhis2/dhis2-core/pull/1453/commits/a1b623680b81399762e4e0aa70230ce1827e911f 
I thought about this approach before but IMHO it makes the code a bit more unreadable. Also as we don't have access to `org.hisp.dhis.query` from dhis-api, therefore, I had to return a `Collection<?>` which is not particularly elegant.

larshelge commented on dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/InterpretationController.java:
```diff
@@ -75,6 +89,56 @@
+            disjunction.add( Restrictions.in( "mentions.username", items ) );
+            disjunction.add( Restrictions.in( "comments.mentions.username", items ) );
+            query.add( disjunction );
+        }
```
119-126: This is too much logic for a controller, can we move to InterpretationService or to MentionUtils?

adrianq replied:
Done here https://github.com/dhis2/dhis2-core/pull/1453/commits/a1b623680b81399762e4e0aa70230ce1827e911f 
I thought about this approach before but IMHO it makes the code a bit more unreadable. Also as we don't have access to `org.hisp.dhis.query` from dhis-api, therefore, I had to return a `Collection<?>` which is not particularly elegant.

larshelge commented on dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/InterpretationController.java:
```diff
@@ -342,10 +406,14 @@ public void updateComment( @PathVariable( "uid" ) String uid, @PathVariable( "cu
+                comment.setMentions(  users );
+                interpretationService.updateSharingForMentions(interpretation, users);
+                interpretationService.updateInterpretation( interpretation );
+                interpretationService.sendNotifications(interpretation, comment, users);
```
Codestyle, space between method and arguments.

adrianq commented:
I have addressed all your comments. I have also merged the latest from master. This branch is not compiling anymore due to some errors in some unrelated tests that are also happening in master branch.

mortenoh commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/interpretation/MentionUtils.java:
```diff
@@ -0,0 +1,53 @@
+
+import org.hisp.dhis.user.User;
+
+public class MentionUtils
```
should use final class and private constructor

adrianq replied:
Done https://github.com/dhis2/dhis2-core/pull/1453/commits/8883c393cdf3cf28338c5ebe133a2355ddac7660

mortenoh commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,26 +213,151 @@ public void deleteInterpretation( Interpretation interpretation )
+    public Set<User> getMentionedUsers( String text )
+    {
+        Set<User> users = new HashSet<>();
+        Matcher matcher = Pattern.compile( "(?:\\s|^)@([\\w+._-]+)" ).matcher( text );
```
Can be made static and externalized, no need to compile every time

adrianq replied:
Done here https://github.com/dhis2/dhis2-core/pull/1453/commits/8c29e1ac66cdf0f3abd2007e6f6c3c28a629c79f

mortenoh commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,26 +213,151 @@ public void deleteInterpretation( Interpretation interpretation )
+    @Override
+    public void sendNotifications( Interpretation interpretation, InterpretationComment comment, Set<User> users )
+    {
+        if ( interpretation != null && users.size() > 0 )
```
wouldn't it be better here to check for `intepretation == null || users.isEmpty()` and then return? now there is an entire indent level added because of this

mortenoh commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,26 +213,151 @@ public void deleteInterpretation( Interpretation interpretation )
+    @Override
+    public List<String> removeCustomFilters( List<String> filters )
+    {
+
```
newline

mortenoh commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,26 +213,151 @@ public void deleteInterpretation( Interpretation interpretation )
+    {
+
+        List<String> mentions = new ArrayList<String>();
+        ListIterator<String> filterIterator = filters.listIterator();
```
newline

mortenoh commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,26 +213,151 @@ public void deleteInterpretation( Interpretation interpretation )
+                mentions.add( filterSplit[2] );
+                filterIterator.remove();
+            }
+        }
```
newline+

mortenoh commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,26 +213,151 @@ public void deleteInterpretation( Interpretation interpretation )
+    @Override
+    public Collection<Disjunction> getDisjunctionsFromCustomMentions( List<String> mentions, Schema schema )
+    {
+
```
newline-

mortenoh commented on dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/dialect/DhisH2Dialect.java:
```diff
@@ -41,6 +41,7 @@
     public DhisH2Dialect()
     {
         registerColumnType( Types.JAVA_OBJECT, "text" );
+        registerColumnType( Types.JAVA_OBJECT, "jsonb" );
```
What is this? did you mean `Types.JAVA_OBJECT-1` ?

adrianq replied:
Yep, you're right. Changed it and apply codestyling https://github.com/dhis2/dhis2-core/pull/1453/commits/87a14361f13f627504c92fe75bcce14a3d519a2a

mortenoh commented on dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/mock/MockUserService.java:
```diff
@@ -0,0 +1,331 @@
+ * @author Adrian Quintana
+ */
+public class MockUserService
+    implements
```
indent

adrianq commented:
@mortenoh I am done with all your comments. I have also extracted another method (removeCustomFilters) to MentionUtils class following the same idea behind some of your comments. Updated with the latest from master

adrianq commented:
Closing in favour of https://github.com/dhis2/dhis2-core/pull/1735
